lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+--
+-- In our current implementation we require the `PointerUTxO` to be created in
+-- the @[0] -> [0,1] transition@, i.e. @nextKeyIx 1 -> nextKeyIx 2@.
+pointerIx
+    :: Int
+    -> Maybe (Index 'Soft 'AddressK)
+pointerIx 0 = Nothing
+pointerIx 1 = Nothing
+pointerIx n = Just $ toEnum n
+
+--------------------------------------------------------------------------------
+-- Constructors
+--------------------------------------------------------------------------------
+
+-- | Construct the initial delegation state.
+initialDelegationState
+    :: k 'AccountK XPub
+    -> DelegationState k
+initialDelegationState accK = DelegationState accK Zero
+
+--------------------------------------------------------------------------------
+-- Chain
+--------------------------------------------------------------------------------
+
+-- | A transaction type specific to `DelegationState`.
+--
+-- Intented to be converted both from and to a more real transaction type.
+--
+-- When constructing a real transaction from `Tx`, these `outputs`
+-- should appear before other outputs. In the theoretical event that there's
+-- also a user-specified output with the same payment key as the pointer output,
+-- `applyTx` will track the first one as the new pointer.
+data Tx = Tx
+    { certs :: [Cert]
+    , inputs :: [(TxIn, Coin)]
+    , outputs :: [TxOut]
+    }
+    deriving (Eq, Generic)
+    deriving Show via (Quiet Tx)
+
+instance Semigroup Tx where
+    (Tx cs1 is1 os1) <> (Tx cs2 is2 os2) =
+        Tx (cs1 <> cs2) (is1 <> is2) (os1 <> os2)
+
+data Cert
+    = RegisterKey RewardAccount
+    | Delegate RewardAccount
+      -- ^ Which pool we're delegating to is here (and for now) irrelevant.
+      -- The main thing is that there exists a witness on-chain for this stake
+      -- key (registration certs don't require witnesses)
+      --
+      -- TODO: We may also want to add the PoolId here.
+    | DeRegisterKey RewardAccount
+    deriving (Eq, Show, Generic)
+
+-- | Given a `DelegationState`, produce a `Tx` registering or de-registering
+-- stake-keys, in order to have @n@ stake-keys.
+--
+-- E.g. @setPortfolioOf s0 _ _ 2@ creates a tx which after application causes
+-- the state to have @activeKeys == [0,1]@
+--
+-- Returns @Nothing@ if the target @n@ is already reached.
+setPortfolioOf
+    :: (SoftDerivation k, ToRewardAccount k)
+    => DelegationState k
+    -> Coin
+        -- ^ minUTxOVal
+    -> (k 'AddressK XPub -> Address)
+        -- ^ A way to construct an Address
+    -> (RewardAccount -> Bool)
+        -- ^ Whether or not the key is registered.
+        --
+        -- TODO: Need a Set or Map for the real implementation with LSQ.
+    -> Int
+        -- ^ Target number of stake keys.
+    -> Maybe Tx
+setPortfolioOf s minUTxOVal mkAddress isReg n =
+    repairKey0IfNeededTx <> changeStateTx
+  where
+    repairKey0IfNeededTx :: Maybe Tx
+    repairKey0IfNeededTx = case repairKey0 $ state s of
```

Comment:
A comment to clarify what this does would be nice. If I understand correctly, this re-register the initial key in case where it has been de-registered by some other mean? Although I do think the code is correct, I am also wondering why this is _necessary_ ? I am generally quite against doing anything on the behalf of users without explicit consent. Imagine the scenario where a user has restored the same wallet in Daedalus and Yoroi. She then de-registers her first stake key in Yoroi which allows it because it doesn't implement this proposal yet. Then, some weeks later while she re-delegate to new pools, the first stake key gets re-registered without her asking anything. 

I don't think the first key should actually be handled so specially. Rather, the multi-stake key state should really only care about keys beyond the first one. If the first key has been de-registered, then so be it :shrug: ... Checking this before trying to de-register it (as you do below) should be sufficient.

Reply:
> Then, some weeks later while she re-delegate to new pools, the first stake key gets re-registered without her asking anything.

`setPortfolioOf n` currently doesn't care about the pool-choice of each key. It couldn't just silently delegate key 0 without knowing what to delegate to. A more concrete scenario:

A user delegates to pools A, B, C with stake keys 0, 1 and 2 respectively.

Yoroi shows her wallet is delegating to A. She undelegates.

According to Future Daedalus she is… in some broken state?… delegating to B, C.

I imagine
- `setPortfolio [A,B,C]` would repair the state by re-registering and delegating key 0
- `setPortfolio [B,C]` would shift `{1 -> B, 2 -> C}` to { 0 -> C, 1 -> B }, by de-registering key 2, and re-registering key 0

To me it seems much easier if we always use key 0 in addresses, some percent of the time, and treat a missing key 0, as a invalid, temporary state. 

When the user un-delegated using the old wallet, the change addresses used key 0. Same for all txs using the old wallet. If the new wallet wants to remove all funds associated with key 0, that's more work.

Reply:
(added a comment)